File: backend/routers/ripng/ripng.py
"""RIPng Protocol Router.

API endpoints for managing VyOS RIPng (Routing Information Protocol next generation)
configuration. Supports version-aware configuration for VyOS 1.4 and 1.5.
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import RipNgBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
import inspect
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/ripng", tags=["ripng"])


# ============================================================================
# Pydantic Models
# ============================================================================


class RipNgDistributeListGlobal(BaseModel):
    access_list_in: Optional[str] = None
    access_list_out: Optional[str] = None
    prefix_list_in: Optional[str] = None
    prefix_list_out: Optional[str] = None


class RipNgDistributeListInterface(BaseModel):
    interface: str
    access_list_in: Optional[str] = None
    access_list_out: Optional[str] = None
    prefix_list_in: Optional[str] = None
    prefix_list_out: Optional[str] = None


class RipNgDistributeList(BaseModel):
    global_filters: RipNgDistributeListGlobal = RipNgDistributeListGlobal()
    interface_filters: List[RipNgDistributeListInterface] = []


class RipNgInterface(BaseModel):
    name: str
    split_horizon: Optional[str] = None


class RipNgRedistribute(BaseModel):
    protocol: str
    metric: Optional[int] = None
    route_map: Optional[str] = None


class RipNgTimers(BaseModel):
    update: Optional[int] = None
    timeout: Optional[int] = None
    garbage_collection: Optional[int] = None


class RipNgConfig(BaseModel):
    default_information_originate: bool = False
    default_metric: Optional[int] = None
    route_map: Optional[str] = None
    aggregate_addresses: List[str] = []
    networks: List[str] = []
    routes: List[str] = []
    passive_interfaces: List[str] = []
    distribute_list: RipNgDistributeList = RipNgDistributeList()
    interfaces: List[RipNgInterface] = []
    redistribute: List[RipNgRedistribute] = []
    timers: RipNgTimers = RipNgTimers()


class RipNgBatchOperation(BaseModel):
    op: str = Field(..., description="Builder method name")
    value: Optional[str] = Field(None, description="Comma-separated arguments")


class RipNgBatchRequest(BaseModel):
    operations: List[RipNgBatchOperation]


class VyOSResponse(BaseModel):
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_ripng_capabilities(request: Request):
    """Return RIPng feature capabilities based on the device VyOS version."""
    await require_read_permission(request, FeatureGroup.RIPNG)

    try:
        service = get_session_vyos_service(request)
        builder = RipNgBatchBuilder(version=service.get_version())
        capabilities = builder.get_capabilities()

        if hasattr(request.state, "instance") and request.state.instance:
            capabilities["instance_name"] = request.state.instance.get("name")
            capabilities["instance_id"] = request.state.instance.get("id")

        return capabilities
    except Exception:
        logger.exception("Unhandled error in get_ripng_capabilities")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 2: Config
# ============================================================================


@router.get("/config", response_model=RipNgConfig)
async def get_ripng_config(http_request: Request, refresh: bool = False):
    """Return the full RIPng configuration in a normalized format."""
    await require_read_permission(http_request, FeatureGroup.RIPNG)

    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        ripng_raw = full_config.get("protocols", {}).get("ripng", {})

        if not ripng_raw:
            return RipNgConfig()

        return RipNgConfig(
            default_information_originate="originate" in (ripng_raw.get("default-information") or {}),
            default_metric=_safe_int(ripng_raw.get("default-metric")),
            route_map=ripng_raw.get("route-map"),
            aggregate_addresses=_to_list(ripng_raw.get("aggregate-address")),
            networks=_to_list(ripng_raw.get("network")),
            routes=_to_list(ripng_raw.get("route")),
            passive_interfaces=_to_list(ripng_raw.get("passive-interface")),
            distribute_list=_parse_distribute_list(ripng_raw.get("distribute-list", {})),
            interfaces=_parse_interfaces(ripng_raw.get("interface", {})),
            redistribute=_parse_redistribute(ripng_raw.get("redistribute", {})),
            timers=_parse_timers(ripng_raw.get("timers", {})),
        )
    except Exception:
        logger.exception("Unhandled error in get_ripng_config")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Config Parsers
# ============================================================================


def _safe_int(value) -> Optional[int]:
    if value is None:
        return None
    try:
        return int(value)
    except (ValueError, TypeError):
        return None


def _to_list(value) -> List[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return value
    if isinstance(value, str):
        return [value]
    if isinstance(value, dict):
        return list(value.keys())
    return []


def _parse_distribute_list(raw: dict) -> RipNgDistributeList:
    if not raw:
        return RipNgDistributeList()

    acl_raw = raw.get("access-list", {}) or {}
    pl_raw = raw.get("prefix-list", {}) or {}

    global_filters = RipNgDistributeListGlobal(
        access_list_in=acl_raw.get("in") if isinstance(acl_raw, dict) else None,
        access_list_out=acl_raw.get("out") if isinstance(acl_raw, dict) else None,
        prefix_list_in=pl_raw.get("in") if isinstance(pl_raw, dict) else None,
        prefix_list_out=pl_raw.get("out") if isinstance(pl_raw, dict) else None,
    )

    iface_filters: List[RipNgDistributeListInterface] = []
    for iface, iface_raw in (raw.get("interface", {}) or {}).items():
        if iface_raw is None:
            iface_raw = {}
        iface_acl = iface_raw.get("access-list", {}) or {}
        iface_pl = iface_raw.get("prefix-list", {}) or {}
        iface_filters.append(RipNgDistributeListInterface(
            interface=iface,
            access_list_in=iface_acl.get("in") if isinstance(iface_acl, dict) else None,
            access_list_out=iface_acl.get("out") if isinstance(iface_acl, dict) else None,
            prefix_list_in=iface_pl.get("in") if isinstance(iface_pl, dict) else None,
            prefix_list_out=iface_pl.get("out") if isinstance(iface_pl, dict) else None,
        ))

    return RipNgDistributeList(global_filters=global_filters, interface_filters=iface_filters)


def _parse_interfaces(raw: dict) -> List[RipNgInterface]:
    if not raw:
        return []

    interfaces = []
    for iface_name, cfg in raw.items():
        if cfg is None:
            cfg = {}

        split_horizon_raw = cfg.get("split-horizon", {}) or {}
        split_horizon = None
        if "poison-reverse" in split_horizon_raw:
            split_horizon = "poison-reverse"
        elif "disable" in split_horizon_raw:
            split_horizon = "disable"

        interfaces.append(RipNgInterface(
            name=iface_name,
            split_horizon=split_horizon,
        ))

    return interfaces


def _parse_redistribute(raw: dict) -> List[RipNgRedistribute]:
    if not raw:
        return []

    entries = []
    for protocol, cfg in raw.items():
        if cfg is None:
            cfg = {}
        entries.append(RipNgRedistribute(
            protocol=protocol,
            metric=_safe_int(cfg.get("metric")),
            route_map=cfg.get("route-map"),
        ))
    return entries


def _parse_timers(raw: dict) -> RipNgTimers:
    if not raw:
        return RipNgTimers()
    return RipNgTimers(
        update=_safe_int(raw.get("update")),
        timeout=_safe_int(raw.get("timeout")),
        garbage_collection=_safe_int(raw.get("garbage-collection")),
    )


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch", response_model=VyOSResponse)
async def ripng_batch_configure(http_request: Request, body: RipNgBatchRequest):
    """Execute a batch of RIPng configuration operations atomically."""
    await require_write_permission(http_request, FeatureGroup.RIPNG)

    try:
        service = get_session_vyos_service(http_request)
        builder = RipNgBatchBuilder(version=service.get_version())

        for operation in body.operations:
            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 0:
                method()
            elif len(params) == 1:
                if operation.value is not None:
                    method(operation.value)
            elif len(params) == 2 and operation.value is not None:
                values = operation.value.split(",", 1)
                if len(values) == 2:
                    method(values[0], values[1])
                else:
                    method(operation.value, "")
            elif len(params) == 3 and operation.value is not None:
                values = operation.value.split(",", 2)
                if len(values) == 3:
                    method(values[0], values[1], values[2])
                elif len(values) == 2:
                    method(values[0], values[1], "")

        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "RIPng configuration updated"},
            error=response.error if response.error else None,
        )
    except AttributeError as e:
        raise HTTPException(status_code=400, detail=f"Unknown operation: {str(e)}")
    except Exception:
        logger.exception("Unhandled error in ripng_batch_configure")
        raise HTTPException(status_code=500, detail="Internal server error")

File: frontend/src/lib/api/ripng.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface RipNgDistributeListGlobal {
  access_list_in?: string | null;
  access_list_out?: string | null;
  prefix_list_in?: string | null;
  prefix_list_out?: string | null;
}

export interface RipNgDistributeListInterface {
  interface: string;
  access_list_in?: string | null;
  access_list_out?: string | null;
  prefix_list_in?: string | null;
  prefix_list_out?: string | null;
}

export interface RipNgDistributeList {
  global_filters: RipNgDistributeListGlobal;
  interface_filters: RipNgDistributeListInterface[];
}

export interface RipNgInterface {
  name: string;
  split_horizon?: string | null;
}

export interface RipNgRedistribute {
  protocol: string;
  metric?: number | null;
  route_map?: string | null;
}

export interface RipNgTimers {
  update?: number | null;
  timeout?: number | null;
  garbage_collection?: number | null;
}

export interface RipNgConfig {
  default_information_originate: boolean;
  default_metric?: number | null;
  route_map?: string | null;
  aggregate_addresses: string[];
  networks: string[];
  routes: string[];
  passive_interfaces: string[];
  distribute_list: RipNgDistributeList;
  interfaces: RipNgInterface[];
  redistribute: RipNgRedistribute[];
  timers: RipNgTimers;
}

export interface RipNgCapabilities {
  version: string;
  version_info: {
    is_1_4: boolean;
    is_1_5: boolean;
  };
  features: {
    global_settings: { supported: boolean; description: string };
    aggregate_addresses: { supported: boolean; description: string };
    networks: { supported: boolean; description: string };
    static_routes: { supported: boolean; description: string };
    passive_interfaces: { supported: boolean; description: string };
    distribute_lists: { supported: boolean; description: string };
    interface_settings: { supported: boolean; description: string };
    redistribute: { supported: boolean; description: string; protocols: string[] };
    timers: { supported: boolean; description: string };
  };
  instance_name?: string;
  instance_id?: string;
}

export interface RipNgBatchOperation {
  op: string;
  value?: string;
}

export interface RipNgBatchRequest {
  operations: RipNgBatchOperation[];
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

// ============================================================================
// API Service
// ============================================================================

class RipNgService {
  async getCapabilities(): Promise<RipNgCapabilities> {
    return apiClient.get<RipNgCapabilities>("/vyos/ripng/capabilities");
  }

  async getConfig(refresh = false): Promise<RipNgConfig> {
    return apiClient.get<RipNgConfig>("/vyos/ripng/config", {
      refresh: refresh.toString(),
    });
  }

  async refreshConfig(): Promise<VyOSResponse> {
    return apiClient.post<VyOSResponse>("/vyos/config/refresh");
  }

  async batchConfigure(request: RipNgBatchRequest): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/ripng/batch", request);
    if (!result.success) throw new Error(result.error || "Operation failed");
    await this.refreshConfig();
    return result;
  }

  // ==========================================================================
  // Global Settings
  // ==========================================================================

  async updateGlobalSettings(
    original: RipNgConfig,
    updated: Partial<RipNgConfig>
  ): Promise<VyOSResponse> {
    const ops: RipNgBatchOperation[] = [];

    if (updated.default_metric !== original.default_metric) {
      ops.push(
        updated.default_metric != null
          ? { op: "set_default_metric", value: String(updated.default_metric) }
          : { op: "delete_default_metric" }
      );
    }

    if (updated.route_map !== original.route_map) {
      ops.push(
        updated.route_map
          ? { op: "set_route_map", value: updated.route_map }
          : { op: "delete_route_map" }
      );
    }

    if (
      updated.default_information_originate !== undefined &&
      updated.default_information_originate !== original.default_information_originate
    ) {
      ops.push(
        updated.default_information_originate
          ? { op: "set_default_information_originate" }
          : { op: "delete_default_information_originate" }
      );
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async updateTimers(
    original: RipNgTimers,
    updated: RipNgTimers
  ): Promise<VyOSResponse> {
    const ops: RipNgBatchOperation[] = [];

    if (updated.update !== original.update) {
      ops.push(
        updated.update != null
          ? { op: "set_timers_update", value: String(updated.update) }
          : { op: "delete_timers_update" }
      );
    }

    if (updated.timeout !== original.timeout) {
      ops.push(
        updated.timeout != null
          ? { op: "set_timers_timeout", value: String(updated.timeout) }
          : { op: "delete_timers_timeout" }
      );
    }

    if (updated.garbage_collection !== original.garbage_collection) {
      ops.push(
        updated.garbage_collection != null
          ? { op: "set_timers_garbage_collection", value: String(updated.garbage_collection) }
          : { op: "delete_timers_garbage_collection" }
      );
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  // ==========================================================================
  // Networks / Aggregate Addresses / Routes / Passive Interfaces
  // ==========================================================================

  async addNetwork(network: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "set_network", value: network }] });
  }

  async removeNetwork(network: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_network", value: network }] });
  }

  async addAggregateAddress(prefix: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "set_aggregate_address", value: prefix }] });
  }

  async removeAggregateAddress(prefix: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_aggregate_address", value: prefix }] });
  }

  async addRoute(prefix: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "set_route", value: prefix }] });
  }

  async removeRoute(prefix: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_route", value: prefix }] });
  }

  async addPassiveInterface(iface: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "set_passive_interface", value: iface }] });
  }

  async removePassiveInterface(iface: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_passive_interface", value: iface }] });
  }

  // ==========================================================================
  // Interface Settings (split-horizon only)
  // ==========================================================================

  async createInterface(config: RipNgInterface): Promise<VyOSResponse> {
    const ops: RipNgBatchOperation[] = [{ op: "set_interface", value: config.name }];
    if (config.split_horizon === "disable") {
      ops.push({ op: "set_interface_split_horizon_disable", value: config.name });
    } else if (config.split_horizon === "poison-reverse") {
      ops.push({ op: "set_interface_split_horizon_poison_reverse", value: config.name });
    }
    return this.batchConfigure({ operations: ops });
  }

  async updateInterface(
    original: RipNgInterface,
    updated: RipNgInterface
  ): Promise<VyOSResponse> {
    const ops: RipNgBatchOperation[] = [];
    const name = original.name;

    if (updated.split_horizon !== original.split_horizon) {
      if (original.split_horizon) {
        ops.push({ op: "delete_interface_split_horizon", value: name });
      }
      if (updated.split_horizon === "disable") {
        ops.push({ op: "set_interface_split_horizon_disable", value: name });
      } else if (updated.split_horizon === "poison-reverse") {
        ops.push({ op: "set_interface_split_horizon_poison_reverse", value: name });
      }
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async deleteInterface(name: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_interface", value: name }] });
  }

  // ==========================================================================
  // Redistribute
  // ==========================================================================

  async createRedistribute(entry: RipNgRedistribute): Promise<VyOSResponse> {
    const ops: RipNgBatchOperation[] = [{ op: "set_redistribute", value: entry.protocol }];
    if (entry.metric != null) {
      ops.push({ op: "set_redistribute_metric", value: `${entry.protocol},${entry.metric}` });
    }
    if (entry.route_map) {
      ops.push({ op: "set_redistribute_route_map", value: `${entry.protocol},${entry.route_map}` });
    }
    return this.batchConfigure({ operations: ops });
  }

  async updateRedistribute(
    original: RipNgRedistribute,
    updated: RipNgRedistribute
  ): Promise<VyOSResponse> {
    const ops: RipNgBatchOperation[] = [];
    const proto = original.protocol;

    if (updated.metric !== original.metric) {
      ops.push(
        updated.metric != null
          ? { op: "set_redistribute_metric", value: `${proto},${updated.metric}` }
          : { op: "set_redistribute", value: proto }
      );
    }

    if (updated.route_map !== original.route_map) {
      if (updated.route_map) {
        ops.push({ op: "set_redistribute_route_map", value: `${proto},${updated.route_map}` });
      } else {
        ops.push({ op: "delete_redistribute", value: proto });
        ops.push({ op: "set_redistribute", value: proto });
        if (updated.metric != null) {
          ops.push({ op: "set_redistribute_metric", value: `${proto},${updated.metric}` });
        }
      }
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async deleteRedistribute(protocol: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_redistribute", value: protocol }] });
  }

  // ==========================================================================
  // Distribute List - Global
  // ==========================================================================

  async updateDistributeListGlobal(
    original: RipNgDistributeListGlobal,
    updated: RipNgDistributeListGlobal
  ): Promise<VyOSResponse> {
    const ops: RipNgBatchOperation[] = [];

    if (updated.access_list_in !== original.access_list_in) {
      ops.push(
        updated.access_list_in
          ? { op: "set_distribute_list_access_list_in", value: updated.access_list_in }
          : { op: "delete_distribute_list_access_list_in" }
      );
    }

    if (updated.access_list_out !== original.access_list_out) {
      ops.push(
        updated.access_list_out
          ? { op: "set_distribute_list_access_list_out", value: updated.access_list_out }
          : { op: "delete_distribute_list_access_list_out" }
      );
    }

    if (updated.prefix_list_in !== original.prefix_list_in) {
      ops.push(
        updated.prefix_list_in
          ? { op: "set_distribute_list_prefix_list_in", value: updated.prefix_list_in }
          : { op: "delete_distribute_list_prefix_list_in" }
      );
    }

    if (updated.prefix_list_out !== original.prefix_list_out) {
      ops.push(
        updated.prefix_list_out
          ? { op: "set_distribute_list_prefix_list_out", value: updated.prefix_list_out }
          : { op: "delete_distribute_list_prefix_list_out" }
      );
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  // ==========================================================================
  // Distribute List - Per Interface
  // ==========================================================================

  async createDistributeListInterface(
    entry: RipNgDistributeListInterface
  ): Promise<VyOSResponse> {
    const ops: RipNgBatchOperation[] = [];
    const iface = entry.interface;
    if (entry.access_list_in) {
      ops.push({ op: "set_distribute_list_interface_access_list_in", value: `${iface},${entry.access_list_in}` });
    }
    if (entry.access_list_out) {
      ops.push({ op: "set_distribute_list_interface_access_list_out", value: `${iface},${entry.access_list_out}` });
    }
    if (entry.prefix_list_in) {
      ops.push({ op: "set_distribute_list_interface_prefix_list_in", value: `${iface},${entry.prefix_list_in}` });
    }
    if (entry.prefix_list_out) {
      ops.push({ op: "set_distribute_list_interface_prefix_list_out", value: `${iface},${entry.prefix_list_out}` });
    }
    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async updateDistributeListInterface(
    original: RipNgDistributeListInterface,
    updated: RipNgDistributeListInterface
  ): Promise<VyOSResponse> {
    const ops: RipNgBatchOperation[] = [];
    const iface = original.interface;

    if (updated.access_list_in !== original.access_list_in) {
      ops.push(
        updated.access_list_in
          ? { op: "set_distribute_list_interface_access_list_in", value: `${iface},${updated.access_list_in}` }
          : { op: "delete_distribute_list_interface_access_list_in", value: iface }
      );
    }

    if (updated.access_list_out !== original.access_list_out) {
      ops.push(
        updated.access_list_out
          ? { op: "set_distribute_list_interface_access_list_out", value: `${iface},${updated.access_list_out}` }
          : { op: "delete_distribute_list_interface_access_list_out", value: iface }
      );
    }

    if (updated.prefix_list_in !== original.prefix_list_in) {
      ops.push(
        updated.prefix_list_in
          ? { op: "set_distribute_list_interface_prefix_list_in", value: `${iface},${updated.prefix_list_in}` }
          : { op: "delete_distribute_list_interface_prefix_list_in", value: iface }
      );
    }

    if (updated.prefix_list_out !== original.prefix_list_out) {
      ops.push(
        updated.prefix_list_out
          ? { op: "set_distribute_list_interface_prefix_list_out", value: `${iface},${updated.prefix_list_out}` }
          : { op: "delete_distribute_list_interface_prefix_list_out", value: iface }
      );
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async deleteDistributeListInterface(iface: string): Promise<VyOSResponse> {
    return this.batchConfigure({
      operations: [{ op: "delete_distribute_list_interface", value: iface }],
    });
  }
}

export const ripNgService = new RipNgService();
